### PR TITLE
Publishable Positions Integration

### PR DESCRIPTION
**Pull request by elizabeth-jimenez**
Dual Merge: 
- [FE PR](https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/2650)
- [Mock PR](https://github.com/MetaPhase-Consulting/mock-fsbid/pull/398)

[TM-4682](https://metaphase.atlassian.net/browse/TM-4682)

[TM-4682]: https://metaphase.atlassian.net/browse/TM-4682?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: talentmap_api/fsbid/services/common.py [hunk 1]
```diff
@@ -302,13 +302,13 @@ def get_individual(uri, query, query_mapping_function, jwt_token, mapping_functi
...
+def send_post_back_office(proc_name, package_name, request_body, request_mapping_function, response_mapping_function, jwt_token, return_code='PV_RETURN_CODE_O'):
     url = f"{BACKOFFICE_CRUD_URL}?procName={proc_name}&packageName={package_name}"
     json_body = request_mapping_function(request_body)
     response = requests.post(url, headers={'JWTAuthorization': jwt_token, 'Content-Type': 'application/json'}, json=json_body).json()
-    # if response is None or (response['PV_RETURN_CODE_O'] and response['PV_RETURN_CODE_O'] is not 0):
-    #     logger.error(f"Fsbid call to '{url}' failed.")
-    #     return None
+    if response is None or (response[return_code] and response[return_code] is not 0):
+        logger.error(f"Fsbid call to '{url}' failed.")
+        return None
```
Comment (scott062) [inline talentmap_api/fsbid/services/common.py:311]: Nice way to handle dynamic response errors - wondering if we should work it into the response mapping functions. So whatever the response, good or bad, it'll be obfuscated away from this common call

File: talentmap_api/fsbid/services/publishable_positions.py [hunk 2]
```diff
@@ -35,7 +35,7 @@ def update_capsule_description(jwt_token, id, description, last_updated_date, up
     '''
     ad_id = jwt.decode(jwt_token, verify=False).get('unique_name')
     uri = f"{PUBLISHABLE_POSITIONS_ROOT}/capsule"
-    url = f"{uri}?pos_seq_num={id}&capsule_descr_txt={description}&update_id={last_updated_date}&update_id={updater_id}&ad_id={ad_id}"
+    url = f"{uri}?pos_seq_num={id}&capsule_descr_txt={description}&last_updated_date={last_updated_date}&update_id={updater_id}&ad_id={ad_id}"
```
Comment (scott062) [inline talentmap_api/fsbid/services/publishable_positions.py:38]: Need to decide if we're using their new function from FSBID or the old version - in theory either should work but matters for permissioning etc

File: talentmap_api/fsbid/services/publishable_positions.py [hunk 3]
```diff
@@ -63,3 +63,191 @@ def convert_capsule_query(query):
...
+    )
+
+def edit_publishable_position_req_mapping(request):
+    return {
+      'PV_API_VERSION_I': '',
+      'PV_AD_ID_I': '',
+      'I_POS_SEQ_NUM': request.get('posSeqNum') or '',
+      'I_PPOS_CAPSULE_DESCR_TXT': request.get('positionDetails') or '',
+      'I_PPOS_LAST_UPDT_USER_ID': request.get('lastUpdatedUserID') or '',
+      'I_PPOS_LAST_UPDT_TMSMP_DT': request.get('lastUpdated') or '',
```
Comment (scott062) [inline talentmap_api/fsbid/services/publishable_positions.py:178]: I think the tmsmp are weird here so need to double check what they expect in dev1.

'YYYYMMDD'
YYYYMMDD HH:MM:SS

File: talentmap_api/fsbid/services/publishable_positions.py [hunk 4]
```diff
@@ -63,3 +63,191 @@ def convert_capsule_query(query):
...
+        **args
+    )
+
+def edit_publishable_position_req_mapping(request):
+    return {
+      'PV_API_VERSION_I': '',
+      'PV_AD_ID_I': '',
+      'I_POS_SEQ_NUM': request.get('posSeqNum') or '',
+      'I_PPOS_CAPSULE_DESCR_TXT': request.get('positionDetails') or '',
+      'I_PPOS_LAST_UPDT_USER_ID': request.get('lastUpdatedUserID') or '',
```
Comment (scott062) [inline talentmap_api/fsbid/services/publishable_positions.py:177]: Last or current user updating - need to check in dev1

File: talentmap_api/fsbid/services/common.py (hunk 1 above)
Comment (elizabeth-jimenez) [reply talentmap_api/fsbid/services/common.py:311]: Yea, good idea!

File: talentmap_api/fsbid/services/publishable_positions.py (hunk 3 above)
Comment (elizabeth-jimenez) [reply talentmap_api/fsbid/services/publishable_positions.py:162]: just sending back what they give - formatting on FE just for display

File: talentmap_api/fsbid/services/publishable_positions.py (hunk 4 above)
Comment (elizabeth-jimenez) [reply talentmap_api/fsbid/services/publishable_positions.py:161]: seemed like last was what they wanted, but defs let's dbl check

File: talentmap_api/fsbid/services/publishable_positions.py (hunk 2 above)
Comment (elizabeth-jimenez) [reply talentmap_api/fsbid/services/publishable_positions.py:38]: 😬

File: talentmap_api/fsbid/services/publishable_positions.py [hunk 5]
```diff
@@ -63,3 +63,175 @@ def convert_capsule_query(query):
...
+        'I_PUBS_CD': request.get('statuses') or '',
+        'I_JC_ID': '',
+        'I_ORDER_BY': '',
+    }
+
+def publishable_positions_res_mapping(data):
+    if data is None or (data['O_RETURN_CODE'] and data['O_RETURN_CODE'] is not 0):
+        logger.error(f"Fsbid call for Publishable Positions failed.")
+        return None
+
```
Comment (scott062) [inline talentmap_api/fsbid/services/publishable_positions.py:110]: Really like this

File: talentmap_api/fsbid/services/publishable_positions.py (hunk 3 above)
Comment (scott062) [reply talentmap_api/fsbid/services/publishable_positions.py:162]: Yeah - pretty confident this is already correct. Ignore me

File: talentmap_api/fsbid/services/publishable_positions.py [hunk 6]
```diff
@@ -63,3 +63,175 @@ def convert_capsule_query(query):
...
+      'I_PPOS_LAST_UPDT_USER_ID': request.get('lastUpdatedUserID') or '',
+      'I_PPOS_LAST_UPDT_TMSMP_DT': request.get('lastUpdated') or '',
+    }
+
+def edit_publishable_position_res_mapping(data):
+    if data is None or (data['O_RETURN_CODE'] and data['O_RETURN_CODE'] is not 0):
+        logger.error(f"Fsbid call for Publishable Positions Edit failed.")
+        return None
+
+    return data
```
Comment (scott062) [inline talentmap_api/fsbid/services/publishable_positions.py:170]: And this is where we just don't do anything with it I assume

File: talentmap_api/fsbid/services/publishable_positions.py [hunk 7]
```diff
@@ -63,3 +63,175 @@ def convert_capsule_query(query):
...
+
+def publishable_positions_filter_req_mapping(request):
+    return {
+        'PV_API_VERSION_I': '',
+        'PV_AD_ID_I': '',
+    }
+
+def publishable_positions_filter_res_mapping(data):
+    if data is None or (data['O_RETURN_CODE'] and data['O_RETURN_CODE'] is not 0):
+        logger.error(f"Fsbid call for Publishable Positions failed.")
```
Comment (scott062) [inline talentmap_api/fsbid/services/publishable_positions.py:197]: Add 'filters' to the error message to distinguish

File: talentmap_api/fsbid/views/publishable_positions.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    def post(self, request):
+        '''
+        Edit Publishable Position
+        '''
+        result = services.edit_publishable_position(request.data, request.META['HTTP_JWT'])
+        if result is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        return Response(result)
```
Comment (scott062) [inline talentmap_api/fsbid/views/publishable_positions.py:52]: I think in other places we do 204 - idk if it really matters but just something to note for dev1 testing in network tab

**Comment by scott062** [comment]
Everything good here already - just the error message for filters and outstanding permissions questions

File: talentmap_api/fsbid/services/publishable_positions.py (hunk 2 above)
Comment (elizabeth-jimenez) [reply talentmap_api/fsbid/services/publishable_positions.py]: Yea, i started to update the old stuff that was here, but then realized it's different enough that we should probably just make new stuff - so this isn't being used for this PR and may have never been hooked up, so should be ok - but i'll dbl check

File: talentmap_api/fsbid/services/publishable_positions.py (hunk 5 above)
Comment (elizabeth-jimenez) [reply talentmap_api/fsbid/services/publishable_positions.py:110]: 😏

File: talentmap_api/fsbid/services/publishable_positions.py (hunk 6 above)
Comment (elizabeth-jimenez) [reply talentmap_api/fsbid/services/publishable_positions.py:170]: Yea, passes off to new Read

**Review by scott062 — APPROVED** [review]

**Pull request merged**